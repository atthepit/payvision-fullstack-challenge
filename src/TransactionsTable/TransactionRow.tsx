import React, { MouseEvent } from "react";
import RowContainer from "./RowContainer";
import Transaction from "./Transaction";
import ColumnText from "./ColumnText";
import DetailsContainer from "./DetailsContainer";
import Details from "./Details";
import DetailsColumn from "./DetailsColumn";
import DetailsProperty from "./DetailsProperty";
import PropertyTitle from "./PropertyTitle";
import {
  mobileHidden,
  capitalize,
  alignRight,
  showMobile,
  hidden
} from "./commonStyles";

interface Props {
  showDetails: boolean;
  handleClick?: (id: string) => void;
  id: string;
  name: string;
  brand: number;
  lastFourDigits: string;
  action: string;
  amount: number;
  currencyCode: string;
  trackingCode: string;
  brandId: number;
  firstSixDigits: string;
  expiryMonth: string;
  expiryYear: string;
  nameSpan: number;
  actionSpan: number;
  amountSpan: number;
  currencyCodeSpan: number;
}

class TransactionRow extends React.Component<Props> {
  public handleTransactionClick = (event: MouseEvent<HTMLElement>) => {
    if (!this.props.handleClick) {
      return;
    }

    // Close Details if it's already open
    if (this.props.showDetails) {
      this.props.handleClick("");
      return;
    }

    // Find element with data-id attribute to get the transaction.id
    const target = event.target as HTMLElement;
    const transactionNode = target.closest("[data-id]") as HTMLElement;
    const id = transactionNode.dataset.id;
    if (!id) {
      return;
    }

    this.props.handleClick(id);
  };
  public render() {
    const { nameSpan, actionSpan, amountSpan, currencyCodeSpan } = this.props;
    return (
      <RowContainer active={this.props.showDetails}>
        <Transaction
          active={this.props.showDetails}
          onClick={this.handleTransactionClick}
          data-id={this.props.id}
          data-testid={this.props.id}
        >
          <ColumnText span={nameSpan}>{this.props.name}</ColumnText>
          <ColumnText className={mobileHidden} span={2}>
            {this.props.brand}
          </ColumnText>
          <ColumnText className={`${mobileHidden}`} span={3}>
            XXXX {this.props.lastFourDigits}
          </ColumnText>
          <ColumnText span={actionSpan} className={capitalize}>
            {this.props.action}
          </ColumnText>
          <ColumnText className={alignRight} span={amountSpan}>
            {this.props.amount.toFixed(2)}
          </ColumnText>
          <ColumnText span={currencyCodeSpan}>
            {this.props.currencyCode}
          </ColumnText>
        </Transaction>
        <DetailsContainer
          active={this.props.showDetails}
          data-hidden={!this.props.showDetails}
          data-testid={`${this.props.id}-details`}
        >
          <Details>
            <DetailsColumn>
              <DetailsProperty>
                <PropertyTitle>ID:</PropertyTitle>
                <span>{this.props.id}</span>
              </DetailsProperty>
              <DetailsProperty>
                <PropertyTitle>Tracking code:</PropertyTitle>
                <span>{this.props.trackingCode}</span>
              </DetailsProperty>
              <DetailsProperty className={showMobile}>
                <PropertyTitle>Brand:</PropertyTitle>
                <span>{this.props.brand}</span>
              </DetailsProperty>
              <DetailsProperty>
                <PropertyTitle>Brand ID:</PropertyTitle>
                <span>{this.props.brandId}</span>
              </DetailsProperty>
            </DetailsColumn>
            <DetailsColumn>
              <DetailsProperty>
                <PropertyTitle>First 6 digits:</PropertyTitle>
                <span>{this.props.firstSixDigits}</span>
              </DetailsProperty>
              <DetailsProperty className={showMobile}>
                <PropertyTitle>Last 4 digits:</PropertyTitle>
                <span>{this.props.lastFourDigits}</span>
              </DetailsProperty>
              <DetailsProperty>
                <PropertyTitle>Expiry month:</PropertyTitle>
                <span>{this.props.expiryMonth}</span>
              </DetailsProperty>
              <DetailsProperty>
                <PropertyTitle>Expiry year:</PropertyTitle>
                <span>{this.props.expiryYear}</span>
              </DetailsProperty>
            </DetailsColumn>
          </Details>
        </DetailsContainer>
        {/* Hidden button for accessibility */}
        <button
          className={hidden}
          onClick={this.handleTransactionClick}
          data-id={this.props.id}
        >
          Toggle details
        </button>
      </RowContainer>
    );
  }
}

export default TransactionRow;
