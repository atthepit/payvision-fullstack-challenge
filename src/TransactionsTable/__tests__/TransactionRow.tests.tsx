import React from "react";
import { render, fireEvent, cleanup } from "react-testing-library";
import TransactionRow from "../TransactionRow";
import { transactions } from "../../../tests/fixtures";

afterEach(cleanup);

test("Handle click", () => {
  const handleClick = jest.fn();
  const transaction = transactions[0];
  const showDetails = false;
  const { getByTestId } = render(
    <TransactionRow
      id={transaction.id}
      name={transaction.card.holderName}
      brand={transaction.brandId}
      lastFourDigits={transaction.card.lastFourDigits}
      action={transaction.action}
      amount={transaction.amount}
      currencyCode={transaction.currencyCode}
      trackingCode={transaction.trackingCode}
      brandId={transaction.brandId}
      firstSixDigits={transaction.card.firstSixDigits}
      expiryMonth={transaction.card.expiryMonth}
      expiryYear={transaction.card.expiryYear}
      showDetails={showDetails}
      handleClick={handleClick}
      nameSpan={1}
      actionSpan={1}
      amountSpan={1}
      currencyCodeSpan={1}
    />
  );

  const row = getByTestId(transaction.id);
  fireEvent.click(row);
  expect(handleClick).toHaveBeenCalledTimes(1);
  expect(handleClick).toHaveBeenCalledWith(transaction.id);
});

test("handle click with empty string if already open", () => {
  const handleClick = jest.fn();
  const transaction = transactions[0];
  const showDetails = true;
  const { getByTestId } = render(
    <TransactionRow
      id={transaction.id}
      name={transaction.card.holderName}
      brand={transaction.brandId}
      lastFourDigits={transaction.card.lastFourDigits}
      action={transaction.action}
      amount={transaction.amount}
      currencyCode={transaction.currencyCode}
      trackingCode={transaction.trackingCode}
      brandId={transaction.brandId}
      firstSixDigits={transaction.card.firstSixDigits}
      expiryMonth={transaction.card.expiryMonth}
      expiryYear={transaction.card.expiryYear}
      showDetails={showDetails}
      handleClick={handleClick}
      nameSpan={1}
      actionSpan={1}
      amountSpan={1}
      currencyCodeSpan={1}
    />
  );

  const row = getByTestId(transaction.id);
  fireEvent.click(row);
  expect(handleClick).toHaveBeenCalledTimes(1);
  expect(handleClick).toHaveBeenCalledWith("");
});

test("show details", () => {
  const handleClick = jest.fn();
  const transaction = transactions[0];
  const showDetails = true;
  const { getByText } = render(
    <TransactionRow
      id={transaction.id}
      name={transaction.card.holderName}
      brand={transaction.brandId}
      lastFourDigits={transaction.card.lastFourDigits}
      action={transaction.action}
      amount={transaction.amount}
      currencyCode={transaction.currencyCode}
      trackingCode={transaction.trackingCode}
      brandId={transaction.brandId}
      firstSixDigits={transaction.card.firstSixDigits}
      expiryMonth={transaction.card.expiryMonth}
      expiryYear={transaction.card.expiryYear}
      showDetails={showDetails}
      handleClick={handleClick}
      nameSpan={1}
      actionSpan={1}
      amountSpan={1}
      currencyCodeSpan={1}
    />
  );

  const element = getByText(transaction.id);
  expect(element).toBeDefined();
});

test("snapshot", () => {
  const handleClick = jest.fn();
  const transaction = transactions[0];
  const showDetails = true;
  const { container } = render(
    <TransactionRow
      id={transaction.id}
      name={transaction.card.holderName}
      brand={transaction.brandId}
      lastFourDigits={transaction.card.lastFourDigits}
      action={transaction.action}
      amount={transaction.amount}
      currencyCode={transaction.currencyCode}
      trackingCode={transaction.trackingCode}
      brandId={transaction.brandId}
      firstSixDigits={transaction.card.firstSixDigits}
      expiryMonth={transaction.card.expiryMonth}
      expiryYear={transaction.card.expiryYear}
      showDetails={showDetails}
      handleClick={handleClick}
      nameSpan={3}
      actionSpan={3}
      amountSpan={2}
      currencyCodeSpan={2}
    />
  );

  expect(container).toMatchSnapshot();
});
