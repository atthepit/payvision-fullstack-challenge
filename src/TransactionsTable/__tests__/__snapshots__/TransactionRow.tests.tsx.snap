// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
.emotion-27 {
  display: none;
}

.emotion-1 {
  -webkit-flex-basis: 14%;
  -ms-flex-preferred-size: 14%;
  flex-basis: 14%;
  margin-left: 2%;
  color: #4a4a4a;
}

@media (max-width:767px) {
  .emotion-1 {
    display: none;
  }
}

.emotion-2 {
  -webkit-flex-basis: 22%;
  -ms-flex-preferred-size: 22%;
  flex-basis: 22%;
  margin-left: 2%;
  color: #4a4a4a;
}

@media (max-width:767px) {
  .emotion-2 {
    display: none;
  }
}

.emotion-25 {
  margin: 0 1rem;
  padding: 1rem 1.5rem;
  border-top: 1px solid #e8ebf3;
}

@media (min-width:768px) {
  .emotion-25 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-flow: row nowrap;
    -ms-flex-flow: row nowrap;
    flex-flow: row nowrap;
    max-height: 16rem;
  }
}

.emotion-15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
}

@media (min-width:768px) {
  .emotion-15 {
    -webkit-flex-basis: 50%;
    -ms-flex-preferred-size: 50%;
    flex-basis: 50%;
  }
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  margin-bottom: 2rem;
}

.emotion-7 {
  color: #213d8f;
  margin-right: 1rem;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  margin-bottom: 2rem;
}

@media (min-width:768px) {
  .emotion-12 {
    display: none;
  }
}

.emotion-28 {
  background-color: #f6f7fb;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  padding: 1rem 0;
  margin: 0 1rem;
  border-top: 1px solid #e8ebf3;
  font-weight: 600;
}

.emotion-6:hover {
  cursor: pointer;
}

.emotion-26 {
  overflow: hidden;
  -webkit-transition: max-height 0.25s ease-in-out;
  transition: max-height 0.25s ease-in-out;
  max-height: 50rem;
}

.emotion-0 {
  -webkit-flex-basis: 22%;
  -ms-flex-preferred-size: 22%;
  flex-basis: 22%;
  margin-left: 2%;
  color: #4a4a4a;
}

.emotion-3 {
  -webkit-flex-basis: 22%;
  -ms-flex-preferred-size: 22%;
  flex-basis: 22%;
  margin-left: 2%;
  color: #4a4a4a;
  text-transform: capitalize;
}

.emotion-4 {
  -webkit-flex-basis: 14%;
  -ms-flex-preferred-size: 14%;
  flex-basis: 14%;
  margin-left: 2%;
  color: #4a4a4a;
  text-align: right;
}

.emotion-5 {
  -webkit-flex-basis: 14%;
  -ms-flex-preferred-size: 14%;
  flex-basis: 14%;
  margin-left: 2%;
  color: #4a4a4a;
}

<div>
  <div
    class="emotion-28"
  >
    <div
      class="emotion-6"
      data-id="7b1c2fef-a001-460b-9d9a-7a6d849b1d5d"
      data-testid="7b1c2fef-a001-460b-9d9a-7a6d849b1d5d"
    >
      <div
        class="emotion-0"
        span="3"
      >
        Meredith Gerlach
      </div>
      <div
        class="emotion-1"
        span="2"
      >
        1060
      </div>
      <div
        class="emotion-2"
        span="3"
      >
        XXXX 
        0005
      </div>
      <div
        class="emotion-3"
        span="3"
      >
        payment
      </div>
      <div
        class="emotion-4"
        span="2"
      >
        100.00
      </div>
      <div
        class="emotion-5"
        span="2"
      >
        USD
      </div>
    </div>
    <div
      class="emotion-26"
      data-hidden="false"
      data-testid="7b1c2fef-a001-460b-9d9a-7a6d849b1d5d-details"
    >
      <div
        class="emotion-25"
      >
        <div
          class="emotion-15"
        >
          <div
            class="emotion-8"
          >
            <span
              class="emotion-7"
            >
              ID:
            </span>
            <span>
              7b1c2fef-a001-460b-9d9a-7a6d849b1d5d
            </span>
          </div>
          <div
            class="emotion-8"
          >
            <span
              class="emotion-7"
            >
              Tracking code:
            </span>
            <span>
              24c22a09-5a94-4552-83de-6128f360b906
            </span>
          </div>
          <div
            class="emotion-12"
          >
            <span
              class="emotion-7"
            >
              Brand:
            </span>
            <span>
              1060
            </span>
          </div>
          <div
            class="emotion-8"
          >
            <span
              class="emotion-7"
            >
              Brand ID:
            </span>
            <span>
              1060
            </span>
          </div>
        </div>
        <div
          class="emotion-15"
        >
          <div
            class="emotion-8"
          >
            <span
              class="emotion-7"
            >
              First 6 digits:
            </span>
            <span>
              378282
            </span>
          </div>
          <div
            class="emotion-12"
          >
            <span
              class="emotion-7"
            >
              Last 4 digits:
            </span>
            <span>
              0005
            </span>
          </div>
          <div
            class="emotion-8"
          >
            <span
              class="emotion-7"
            >
              Expiry month:
            </span>
            <span>
              03
            </span>
          </div>
          <div
            class="emotion-8"
          >
            <span
              class="emotion-7"
            >
              Expiry year:
            </span>
            <span>
              2020
            </span>
          </div>
        </div>
      </div>
    </div>
    <button
      class="emotion-27"
      data-id="7b1c2fef-a001-460b-9d9a-7a6d849b1d5d"
    >
      Toggle details
    </button>
  </div>
</div>
`;
